src/lammpsio/topology.py

```diff
@@ -0,0 +1,171 @@
+            if v.shape != (self.N,):
+                raise TypeError("typeids must be a size N array")
+            if not self.has_typeid():
+                self._typeid = numpy.ones(self.N, dtype=int)
+            numpy.copyto(self._typeid, v)
+        else:
+            self._typeid = None
+
+    def has_typeid(self):
+        """Check if configuration has connection typeids.
+
+        Returns
+        -------
+        bool
+            True if connection typeids have been initialized.
+
+        """
+        return self._typeid is not None
+
+    @property
+    def members(self):
+        """:class:`numpy.ndarray`: Connection members."""
+        if not self.has_members():
+            self._members = numpy.ones((self.N, self._num_members), dtype=int)
+        return self._members
+
+    @members.setter
+    def members(self, value):
+        if value is not None:
+            v = numpy.array(value, ndmin=2, copy=False, dtype=int)
+            if v.shape != (self.N, self._num_members):
+                raise TypeError("Members must be a size N x number of arrays array")
+            numpy.copyto(self._members, v)
+        else:
+            self._members = None
+
+    def has_members(self):
+        """Check if configuration has connection members.
+
+        Returns
+        -------
+        bool
+            True if particle members have been initialized.
+
+        """
+        return self._members is not None
+
+    @property
+    def num_types(self):
+        """:class:`Int`: num_types."""
+        if self._num_types is not None:
+            return self._num_types
+        else:
+            if self.has_typeid():
+                return numpy.amax(numpy.unique(self.typeid))
+            else:
+                return 1
+
+    @num_types.setter
+    def num_types(self, value):
+        if value is not None:
+            self._num_types = int(value)
+        else:
+            self._num_types = None
+
+    def reorder(self, order, check_order=True):
+        """Reorder the connections in place.
+
+        Parameters
+        ----------
+        order : list
+            New order of indexes.
+        check_order : bool
+            If true, validate the new ``order`` before applying it.
+
+        """
+        # sanity check the sorting order before applying it
+        if check_order and self.N > 1:
+            sorted_order = numpy.sort(order)
+            if (
+                sorted_order[0] != 0
+                or sorted_order[-1] != self.N - 1
+                or not numpy.all(sorted_order[1:] == sorted_order[:-1] + 1)
+            ):
+                raise ValueError("New order must be an array from 0 to N-1")
+
+        if self.has_id():
+            self._id = self._id[order]
+        if self.has_typeid():
+            self._typeid = self._typeid[order]
+        if self.has_members():
+            self._members = self._members[order]
+
+
+class Bonds(Topology):
+    def __init__(self, N, num_types):
```

Comment:
We should allow the default value here:
```suggestion
    def __init__(self, N, num_types=None):
```

Reply:
For future reference, can you remind me when it is acceptable to put default arguments in derived classes / not? I remember talking about it once, but I don't recall the specifics

Reply:
It is always allowed by normal Python standards. However, in relentless operations, we don't use default arguments in operations for specific simulation engines to make sure we don't accidentally miss an argument in the wrapper operation.